### ISSUE TITLE
When switch reinstall compiler compile fails, switch can't be fully recovered

### ISSUE THREAD

**Issue by dsheets** [ocaml/opam#1807]
I'm on 60e114a (after #1728) and I had a switch reinstall fail broken. The compiler failed to build and then re-creation of the package state was totally broken with errors like:

```
[ERROR] os is not a valid variable.
[WARNING] Invalid variable os in filter
```

```
[ERROR] make is not a valid variable.
[WARNING] Invalid variable make in filter
```

```
[ERROR] prefix is not a valid variable.
```

Also, all the previous state from the switch (e.g., in `share`) seems to be missing. This doesn't really bother me but it seems inconsistent with the refusal to delete non-empty share directories on package removal.

Perhaps resolving this issue is simply a matter of saving a little more state per switch or outputting a clearer error message (or catching a failed compiler rebuild).

**Comment by AltGr** [comment]
`switch reinstall` is quite brutal: it'll move all contents of the prefix away, then reinstall from scratch. This is good for recovery and cleanup in some cases, but may not be what you want.

The errors on broken semi-installed switches above should have been fixed recently though, so closing this. Feel free to open a new issue if a lighter version of `reinstall` seems useful or if you think this should be better documented. With compilers-as-packages, that version could be `opam reinstall ocaml`.
